gazebo_plugins/src/gazebo_ros_vacuum_gripper.cpp

```diff
@@ -0,0 +1,247 @@
+    return;
+  }
+
+  rosnode_ = new ros::NodeHandle(robot_namespace_);
+
+  // Custom Callback Queue
+  ros::AdvertiseOptions ao = ros::AdvertiseOptions::create<std_msgs::Bool>(
+    topic_name_, 1,
+    boost::bind(&GazeboRosVacuumGripper::Connect, this),
+    boost::bind(&GazeboRosVacuumGripper::Disconnect, this),
+    ros::VoidPtr(), &queue_);
+  pub_ = rosnode_->advertise(ao);
+
+  // Custom Callback Queue
+  ros::AdvertiseServiceOptions aso1 =
+    ros::AdvertiseServiceOptions::create<std_srvs::Empty>(
+    "on", boost::bind(&GazeboRosVacuumGripper::OnServiceCallback,
+    this, _1, _2), ros::VoidPtr(), &queue_);
+  srv1_ = rosnode_->advertiseService(aso1);
+  ros::AdvertiseServiceOptions aso2 =
+    ros::AdvertiseServiceOptions::create<std_srvs::Empty>(
+    "off", boost::bind(&GazeboRosVacuumGripper::OffServiceCallback,
+    this, _1, _2), ros::VoidPtr(), &queue_);
+  srv2_ = rosnode_->advertiseService(aso2);
+
+  // Custom Callback Queue
+  callback_queue_thread_ = boost::thread( boost::bind( &GazeboRosVacuumGripper::QueueThread,this ) );
+
+  // New Mechanism for Updating every World Cycle
+  // Listen to the update event. This event is broadcast every
+  // simulation iteration.
+  update_connection_ = event::Events::ConnectWorldUpdateBegin(
+      boost::bind(&GazeboRosVacuumGripper::UpdateChild, this));
+
+  ROS_INFO("Loaded gazebo_ros_vacuum_gripper");
+}
+
+bool GazeboRosVacuumGripper::OnServiceCallback(std_srvs::Empty::Request &req,
+                                     std_srvs::Empty::Response &res)
+{
+  if (status_) {
+    ROS_WARN("gazebo_ros_vacuum_gripper: already status is 'on'");
+  } else {
+    status_ = true;
+    ROS_INFO("gazebo_ros_vacuum_gripper: status: off -> on");
+  }
+  return true;
+}
+bool GazeboRosVacuumGripper::OffServiceCallback(std_srvs::Empty::Request &req,
+                                     std_srvs::Empty::Response &res)
+{
+  if (status_) {
+    status_ = false;
+    ROS_INFO("gazebo_ros_vacuum_gripper: status: on -> off");
+  } else {
+    ROS_WARN("gazebo_ros_vacuum_gripper: already status is 'off'");
+  }
+  return true;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Update the controller
+void GazeboRosVacuumGripper::UpdateChild()
+{
+  std_msgs::Bool grasping_msg;
+  grasping_msg.data = false;
+  if (!status_) {
+    pub_.publish(grasping_msg);
+    return;
+  }
+  // apply force
+  lock_.lock();
+  math::Pose parent_pose = link_->GetWorldPose();
+  physics::Model_V models = world_->GetModels();
+  for (size_t i = 0; i < models.size(); i++) {
+    if (models[i]->GetName() == link_->GetName() ||
+        models[i]->GetName() == parent_->GetName())
+    {
+      continue;
+    }
+    physics::Link_V links = models[i]->GetLinks();
+    for (size_t j = 0; j < links.size(); j++) {
+      math::Pose link_pose = links[j]->GetWorldPose();
+      math::Pose diff = parent_pose - link_pose;
```

Comment:
This uses the difference in link pose between the end effector and an object to determine whether they are close enough for force to be applied, but this depends on where the coordinate frames are defined. This might work for small objects, but it would be better to use collision checking (such as ray-casting) to determine the proximity to the actual surface, which won't be affected by how coordinate frames are defined.

Let me know if that didn't make sense and we can discuss further.